<!DOCTYPE html>
<html>
<head>
	<meta charset="utf-8">
	<title>LinkedCharts - Supplement to the paper</title>
	<script type="text/javascript" src="src/linked-charts.min.js"></script>
	<script type="text/javascript" src="src/prism.js"></script>
	<link rel="stylesheet" type="text/css" href="src/prism.css">
	<link rel="stylesheet" type="text/css" href="src/linked-charts.css">
	<link rel="stylesheet" type="text/css" href="src/style.css">
</head>
<body onload="setNavHeight();">
	<div id="main">
		<div id="header">
			<h1>LinkedCharts - Supplement to the paper</h1>
			<p>This page provides live versions of the figures from the main paper. Each example is given as a <i>full</i> version with all the decorations, labels and titles and as a <i>minimalistic</i> app with only essential features.</p> 
			<p>For each example, one can look at the full code requried to generate the app both as an R script and as a JavaScript code. To run the R code, the "rlc" package must be installed from <a href="https://CRAN.R-project.org/package=rlc">CRAN</a>, or <a href="https://github.com/anders-biostat/rlc">GitHub</a> for the most recent version. The R code can also be run in Jupyter notebooks (with the IRkernel installed); however, since such R sessions are not considered interactive, one must explicitly tell the R session to wait for the messages from the app. To this end, the "rlc" package provides the "listen" function. The JavaScript code relies on the <i>linked-charts.js</i> library that can be downloaded from <a href="https://anders-biostat.github.io/linked-charts/js/">here</a>. Links to all the required data for each example are given in the description below.</p>
		</div>
		<div id="gallery">
			<div id="navigation">
				<div id="tabs"></div>
			</div>
			<div id="panel"></div>
			<div id="example">
				<div class="aspect-ratio">
					<iframe frameBorder="0"></iframe>	
				</div>
			</div>
			<div id="description"></div>
		</div>		
	</div>
	<script type="text/javascript">
		const examples = [{
				name: "1. Oral cancer, differential expression",
				id: "oc_de",
				ar: "45%",
				descr: 'An example for two linked charts, based on a study by <a href="https://doi.org/10.18632/oncotarget.5529">Conway et al. (2015)</a> comparing cancerous and normal tissues from 19 patients. The MA plot (left) shows all genes with their average expression on the X-axis and log<sub>2</sub>-fold change between tumour and notmal on the Y-axis. Red indicates genes where the difference was reported as significant by the "voom" function of the <href="https://bioconductor.org/packages/release/bioc/html/limma.html">"limma"</a> package. The plot to the right shows expression values (counts per million, CPM) for a selected gene and all the samples. When the user clicks on a point of the MA plot, the expression plot changes, showing the new selected gene. Thus, one can rapidly gain an impression of the details hidden in a summarizing plot like the MA plot.<br><br>To run the provided code, please, download the following data:<br><b>For R:</b> <a href="https://anders-biostat.github.io/linked-charts/rlc/tutorials/oscc/oscc.rda" target="_blank">ocss.rda</a><br><b>For JS:</b> <a target="_blank" href="https://github.com/anders-biostat/linked-charts/raw/master/docs/rlc/tutorials/oscc/maData.js">maData.js</a> and <a target="_blank" href="https://github.com/anders-biostat/linked-charts/raw/master/docs/rlc/tutorials/oscc/heatmapData.js">heatmapData.js</a>'
			}, {
				name: "2. Basic syntax, Iris dataset",
				id: "basic_synt",
				ar: "45%",
				descr: 'This very basic example shows how to make a simple "non-linked" plot with LinkedCharts. Even though the linking functionality is not used here, there is still some built-in interactivity in every chart made with the LinkedCharts library, such as zooming, panning and saving the current state of the plot as an image. Try the tools panel (opens by the grey button in the upper right corner) to explore all possible actions.<br><br>To run the provided code, download the following data:<br><b>For R:</b> The Iris dataset can be loaded with the \'data("iris")\' command<br><b>For JS:</b> <a target="_blank" href="https://github.com/kloivenn/kloivenn.github.io/raw/master/Bioconductor2018/src/js/iris.js">iris.js</a>'
			}, {
				name: "3. Example gallery",
				id: "gallery",
				ar: "57%",
				descr: 'Here, one can find a collection of all charts of the R/LinkedCharts library ("rlc" package). Most examples use randomly generated data simply for demonstration purposes (the code to generate the data is also provided). Some of them use the Iris dataset as an example (<a target="_blank" href="examples/gallery/iris.js">iris.js</a> or run \'data("iris")\' in your R session). Since all these examples are quite basic, there is no minimal code version.',
				no_min: true
			}, {
				name: "4. Drug screening pipeline",
				id: "drug",
				ar: "45%",
				descr: 'This example is based on a drug screening experiment from <a href="https://dx.doi.org/10.1007%2F978-1-4939-7493-1_17">He et al. (2018)</a> and represents a typical analysis pipeline in such an experiment. The heatmap shows correlation values for 50 drugs tested against 21 pancreatic cancer cell lines. One can select a pair of drugs by clicking on a corresponding heatmap cell. Then, the scatter plot will be updated to display individual drug sensitivity scores of all tested cell lines for the two selected drugs against each other, thus providing a way to explore the data behind the correlation value shown in the heatmap. A click on a point in the scatter plot will select a cell line. Then, the other scatter plot below will show the data used to calculate the Drug Sensitivity Score, which is based on the area under the sigmoid curve fitted to viability values for all the tested concentrations.  At the same time, the raw read-out from the plates that contain the selected cell line and drugs are shown to the right as two heatmaps. Moving the mouse of a point in the dose-response plots highlights the corresponding well.<br><br>To run the provided code, please, download the following data:<br><b>For R:</b> <a target="_blank" href="https://github.com/anders-biostat/linked-charts/raw/master/docs/thesis/drugScreen.RData">drugScreen.RData</a><br><b>For JS:</b> <a target="_blank" href="examples/drug/rawData.js">rawData.js</a> and <a target="_blank" href="examples/drug/scoreData.js">scoreData.js</a>'
			}, {
				name: "5. Oral cancer, sample correlation",
				id: "oc_sc",
				ar: "45%",
				descr: 'The heatmap (C) shows Spearman correlation values of gene expression in cancerous, normal and dysplasic tissues from 19 patients (<a href="https://doi.org/10.18632/oncotarget.5529">Conway et al. (2015)</a>). Clicking on any cell of the heatmap, one can explore the expression patterns in the two samples compared there: the scatter plot (D) will show gene expression values for the two selected samples against each other. In this way, it is easy to get some feeling of the data behind each correlation value and check, for example, what abnormalities in the data caused the outliers (in the right bottom corner of the correlation heatmap) to have lower correlation values than all other samples.<br><br>To run the provided code, please, download the following data:<br><b>For R:</b> <a href="https://anders-biostat.github.io/linked-charts/rlc/tutorials/oscc/oscc.rda" target="_blank">ocss.rda</a><br><b>For JS:</b> <a target="_blank" href="https://github.com/anders-biostat/linked-charts/raw/master/docs/rlc/tutorials/oscc/maData.js">maData.js</a> and <a target="_blank" href="https://github.com/anders-biostat/linked-charts/raw/master/docs/rlc/tutorials/oscc/heatmapData.js">heatmapData.js</a>'
			}, {
				name: "+",
				id: "extra",
				ar: "73%",
				descr: 'Here you can find a list of several additional examples of the LinkedCharts usage, most of which are accomanied by detailed tutorials. Click the correspoing linked-charts to learn more about each of the mentioned examples.'
			}];

		function setNavHeight() {
			d3.select("#navigation")
				.style("height", (d3.select(".tab").node().getBoundingClientRect().height - 1) + "px");
		}

		var code = "full", cur_ex = 0;

		d3.select("#tabs").selectAll(".tab")
			.data(examples)
			.enter()
			.append("div")
				.attr("class", (d, i) => i == 0 ? "tab active" : "tab")
				.text(d => d.name)
				.style("border-left", (d, i) => i == 0 ? "1px #224 solid" : "none")
				.on("click", function(d, i) {
					cur_ex = i;
					d3.select("#tabs").selectAll(".tab").classed("active", false);
					d3.select(this).classed("active", true);
					showExample(d);
				});
		d3.select("#navigation")
			.append("div")
				.attr("class", "filler");

		d3.select("#panel").selectAll(".panel-block")
			.data(["version", "code"])
			.enter()
			.append("div")
				.attr("class", "panel-block")
				.attr("id", d => d)
				.append("div")
					.attr("class", "label")
					.text(d => d == "version" ? "Version:" : "See code:");

		d3.select("#panel")
			.selectAll(".panel-block").selectAll(".button")
			.data(d => d == "version" ? ["Full", "Minimal"] : ["R", "JS"])
			.enter()
			.append("div")
				.attr("class", "button")
				.attr("id", d => d == "Full" ? "full" : d == "Minimal" ? "min" : d == "R" ? "rcode" : "jscode")
				.text(d => d == "Full" ? "full" : d == "Minimal" ? "minimal" : d == "R" ? "R" : "JavaScript");

		d3.select("#panel").select("#" + code)
			.classed("active", true);

		d3.select("#panel").selectAll("#full,#min")
			.on("click", switchCode);
		d3.select("#panel").selectAll("#rcode,#jscode")
			.on("click", showCode);

		showExample(examples[0]);
		
		function showExample(d) {
			if(d.name == "+") {
				d3.select("#panel")
					.style("display", "none");
			} else {
				d3.select("#panel")
					.style("display", "block");				
			}
			d3.select("#example").select(".aspect-ratio")
				.style("padding-bottom", d.ar);
			if(d.no_min) {
				code = "full";
				d3.select("#full")
					.classed("active", true);
				d3.select("#min")
					.classed("disabled", true)
					.on("click", null)
			} else {
				d3.select("#min")
					.classed("disabled", false)
					.on("click", switchCode)			
			}
			d3.select("#example").select("iframe")
				.attr("src", "examples/" + d.id + "/" + d.id + "_" + code + ".html");
			d3.select("#description")
				.html(d.descr);
		}

		function switchCode(d) {
			d3.select("#panel").selectAll(".button")
				.classed("active", false);
			d3.select(this)
				.classed("active", true);
			code = this.id;
			d3.select("#example").select("iframe")
				.attr("src", "examples/" + examples[cur_ex].id + "/" + examples[cur_ex].id + "_" + code + ".html");
		}

		function showCode(d) {
			const lang = d.split(" ")[0];
			window.open("examples/" + examples[cur_ex].id + "/" + lang + "_code_" + code + ".html", 
				lang + " code for the example", "width=800,height=600");
		}
	</script>
</body>
</html>